### DISCUSSION TITLE
Modals not dismissing on forward routing

### DISCUSSION THREAD

**Discussion by TylerK** [vercel/next.js#50284] [App Router]
Following along with the examples given in the [Parallel Routing](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes) docs, I have the following route structure:

```
app
├─ @authModal
│  ├─ (.)auth
│  │  └─ login
│  │  │  └─ page.tsx // Contains the modal with login form 
│  └─ [...catchAll]
│  │  │  └─ page.tsx 
├─ dashboard
│  └─ page.tsx
├─ auth
│  └─ login
│  │  └─ page.tsx // Contains only the login form
```

Almost everything works beautifully. 

✅ Clicking Login via `<Link href="/auth/login" />` on the homepage will pop open the login modal
✅ Router will reflect `/auth/login`
✅ Pressing refresh will render `/auth/login` as a page, as expected
❌ Navigating via `router.push('/dashboard')` from within the modal does not close the modal

The docs seem to indicate this setup should work:

```
If you want to navigate elsewhere and dismiss a modal, you can also use a catch-all route.
```

This doesn't appear to be working. Any help would be greatly appreciated!

**Comment by TylerK** [comment]
Found a workaround I don't necessarily love: 

```jsx
// AuthModal.tsx
import { usePathname } from 'next/navigation';

export function AuthModal({ children }: { children: ReactNode }) {
    const pathname = usePathname();
    if (!pathname.includes('auth')) {
        return null;
    }
    
    // ... 
}
```

If anyone has a better way of handling preventing the intercepted Modal from remaining on screen after a `router.push()` event I'm all ears :)

**Reply by charlesloubao** [reply]
Another option is to use `window.location.replace()` but you lose the current context

**Reply by TylerK** [reply]
Yeah I'm going with this for now, it's ugly but it "works" :)

**Reply by olegvlotus** [reply]
Thank you for this workaround, that works until a fix is found!

**Reply by wlechowicz** [reply]
This is still the only fix that makes sense, because adding extra stuff to the parallel route with pages and returning null from those doesn't work for routes with dynamic segments as explained here: https://github.com/vercel/next.js/discussions/50284#discussioncomment-7482282

**Reply by isaaccarrington** [reply]
In 2025 this is still the answer I have landed on today :D

**Comment by hussam-001** [comment]
temp solution

```tsx
router.push("/dashboard", {
    // TODO: remove this when [...catchAll] works
    forceOptimisticNavigation: true,
});
```

**Reply by TylerK** [reply]
Thanks Hussam, unfortunately this doesn't appear to work on 13.4

**Reply by hussam-001** [reply]
so weird bcz im also using 13.4

**Reply by kasir-barati** [reply]
Not gonna work when you are import it like this: `import { useRouter } from 'next/navigation';`

I also tried:
```ts
router.push('/somewhere');
router.refresh();
```
to no avail :/.

**Comment by olegvlotus** [comment]
Have the same issue, modal doesn't dismiss when navigating to other page with [...catchAll]

**Comment by hoangminhtran94** [comment]
Still having the same problem in next 13.4.19

**Comment by mastoj** [comment]
This was driving me crazy since I thought I did something wrong. Is it a bug that `router.push` does not dismiss the current intercepted route or is it by design? If it is by design how are we supposed to dismiss modals?

**Reply by alexander-rebello** [reply]
I'm not sure, if it's a bug or supposed to work like this, but the docs recommend using `router.back()` to dismiss modals.

Have a look at [https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#closing-the-modal](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#closing-the-modal)

**Reply by wlechowicz** [reply]
`router.back()` only works in the most simple scenario, it stops working as soon as your modal has links to other intercepted routes.
For example, I have lists of movie thumbnails on the home page (think Netflix), each one opens up a modal with movie details (like /movies/:id). If the movie details page has recommendations for other movies, that's a link to an intercepted route inside an intercepted route, so if you go from home page to movie A details modal (/movies/A), from there to movie B details modal through recs (/movies/B), then `router.back()` does not dismiss the modal but goes back to movie A (/movies/A).

**Comment by gragland** [comment]
The issue is that `[...catchAll]` isn't working for some reason. You can work around this by duplicating your page structure within the parallel route (or at least the pages you want to allow navigation to). Make each of these duplicate pages return null so that they get matched and the modal unmounts. For example:
```
app
├─ @authModal
│  ├─ (.)auth
│  │  └─ login
│  │  │  └─ page.tsx
│  └─ dashboard
│  │  │  └─ page.tsx (this should return null)
├─ dashboard
│  └─ page.tsx
├─ auth
│  └─ login
│  │  └─ page.tsx
```
The modal will now disappear when navigating to `/dashboard`.

**Reply by mastoj** [reply]
I guess you mean to write 

```
app
├─ @authModal
│  ├─ (.)auth
│  │  └─ login
│  │  │  └─ page.tsx
│  └─ dashboard
│  │  │  └─ page.tsx (this should return null)
├─ dashboard
│  └─ page.tsx
├─ auth
│  └─ login
│  │  └─ page.tsx
```

or did you intentionally use `[dashboard]`?

**Reply by gragland** [reply]
@mastoj thanks, fixed :)

**Reply by wlechowicz** [reply]
The problem with this workaround is that if `dashboard` has any dynamic segments inside, you're screwed. Adding `@authModal/dashboard/[...catchAll]/page.tsx` that returns null also doesn't work, just like the `@authModal/[...catchAll]/page.tsx` that's supposed to work according to docs. Is it that no catchAlls work under a parallel route?

**Comment by benyarb** [comment]
Having the same issue in 13.5.3

**Comment by Flying0101** [comment]
same issue.. nothing seems to work properly for me, using 13.5.6 version.. 

router.push("/dashboard", {
    // TODO: remove this when [...catchAll] works
    forceOptimisticNavigation: true,
});

this works well the first time i navigate to another route, but if I open the modal again and try going to another route(or the same) the modal does not close.

**Comment by wlechowicz** [comment]
Unfortunately I ran into the same issue and spent way too much time thinking I'm doing something wrong. I followed the docs for parallel and intercepting routes to set up a basic modal and I cannot get it to dismiss with either a catchAll or router.push()
It only dismisses through router.back() which is silly if you have Link in your modal and they take you to other, non-intercepted pages.

Verified the behavior to be the same on Next 13.5.6 up to 14.0.2 canary.

The official example linked in the docs - "Nextgram" - kinda works because it doesn't let you do anything once the modal is open and there are no other pages in the app, beside "photos".

**Comment by wlechowicz** [comment]
This problem should be fixed with #58215 unless your catch-all is under a dynamic segment, then #58272

**Reply by Enkratia** [reply]
@wlechowicz, it's not fixed. It doesn't work(unmount) without author's workaround.

**Reply by wlechowicz** [reply]
@Enkratia hey man, what can I say. From my experience, those things get fixed (on purpose or by accident) in one canary and then broken in the next one, I stopped trying to keep up with every latest canary and retesting this because it's just too unstable. I'm waiting for a focused effort that fixes all the big known bugs with AppRouter before considering it production ready or wasting more time on it.

**Reply by Enkratia** [reply]
@wlechowicz, got it,  thanks for your efforts

**Comment by Neo-Ciber94** [comment]
My current solution to close all modals is using `useSelectedLayoutSegment`, in my use case it works not sure if is the same for others:

```ts
  const segments = useSelectedLayoutSegment();

  const backCount = segments == null ? 1 : segments.split('/').length + 1;
  for (let i = 0; i < backCount; i++) {
    router.back();
  }
```

**Comment by MrOxMasTer** [comment]
[...catchAll] still doesn't work. Version: 14.1.1 canary 36

**Comment by darthmaim** [comment]
This is still not fixed in the latest canary (`14.2.0-canary.40`).

Is there already an issue with a reproduction for this?

For now I'm using this in my modal, which seems to work for my simple usecase:

```ts
const pathname = usePathname();
const initialPath = useRef(pathname);

if(pathname !== initialPath.current) {
  return null;
}
```

**Reply by wlechowicz** [reply]
Still hit and miss for me, For example, on 14.1.3 the modal doesn't work at all when using `next start`, the links just open the page and it doesn't get intercepted. It works in dev, but doesn't in production - lovely.

**Comment by laduniestu** [comment]
still not work in (14.1.4)
```js
const onSubmit = (values: z.infer<typeof GuestSchema>) => {
    setError('')
    startTransition(() => {
      editGuest(values, id)
        .then((data) => {
          if (data?.error) {
            setError(data.error)
          }
          if (data?.success) {
            toast.success(data.success)
            router.back()
          }
        })
    })
  }
```
when success submitting the form, the `route.back()` is triggered but the modal not close

https://github.com/vercel/next.js/assets/44757637/c4e5cd46-a2f8-4bc7-bb1d-1af41abe94b1


but when click the close button which have function `()=>router.back()` the modal closed.

```js
<Button variant={'outline'} type='button' onClick={() => router.back()}>Close</Button>
```

https://github.com/vercel/next.js/assets/44757637/4e73d2ea-b379-41b4-992c-25b52714d8f6


i have already tried to use `window.location.replace()`
```js
const onSubmit = (values: z.infer<typeof GuestSchema>) => {
    setError('')
    startTransition(() => {
      editGuest(values, id)
        .then((data) => {
          if (data?.error) {
            setError(data.error)
          }
          if (data?.success) {
            toast.success(data.success)
            window.location.replace("/admin/guestlist")
          }
        })
    })
  }
  ```
  this is work but it will reset all the context
  

https://github.com/vercel/next.js/assets/44757637/8c777e50-aa9b-430b-bda0-161c4d806b48

**Reply by siphosenkosindhlovu** [reply]
Have the same issue, but I get a client side exception when I navigate backwards with `router.back()`

**Reply by sejoalfaro** [reply]
### Description
I have similar behavior. In my case I am rendering 2 nested layers as modals. When I try to use server actions to submit the second modal form, router.back doesn't work and something doesn't allow me to touch anything on the screen. So to solve it I am using a route handler just to cover this case.

#### With Route Handlers
https://github.com/vercel/next.js/assets/24913036/c125d4e5-a919-4b04-9d81-95825e5fe4ac

#### With Server Actions
https://github.com/vercel/next.js/assets/24913036/b860bcb1-b422-4d34-b5c1-5d53e3942812

**Comment by proplayercoding** [comment]
i use this one
```
   const onSubmit = () => {
      router.replace(`/dashboard`)
      router.refresh()
  }
```

**Reply by kasir-barati** [reply]
Did not work in my scenario :smiling_face_with_tear:.

I am trying to redirect user to another URL, but it is not necessarily a new URL. It can be the same URL or a completely different one.

```ts
import { useRouter } from 'next/navigation';

// ...
const router = useRouter();
router.push(newPath); // newPath is passed down as an arg
router.refresh();
// ...
```

**Comment by qFlensT** [comment]
Faced the same problem and wasted a lot of time. That's the main problem with Next JS - you don't realize when you're doing something wrong, and when it's just a bug and it doesn't work as it should 

**Updated**
using this:
```ts
export const FixModalCloseBug = ({
  expectedPath,
  children,
}: {
  expectedPath: string | RegExp;
  children: ReactNode;
}) => {
  const pathname = usePathname();

  if (expectedPath instanceof RegExp) {
    if (expectedPath.test(pathname)) {
      return children;
    }
    return null;
  } else if (pathname.includes(expectedPath)) {
    return children;
  }
  return null;
};
```

Then simply:
```ts
export default function Page({ params }: { params: any }) {
  return (
    <FixModalCloseBug expectedPath={"/subscription"}>
      <SubscriptionModal params={params} />
    </FixModalCloseBug>
  );
}
```

**Reply by jiaweing** [reply]
Exactly this.

**Reply by ArturLinnik** [reply]
Thank you for this

**Reply by heron2014** [reply]
Thanks, this seemed to work for me too!

**Comment by raiyansarker** [comment]
I hope this might help, I didn't find any other good solution,
```ts
"use client";

import { Sheet, SheetContent } from "@/components/ui/sheet";
import { usePathname, useRouter } from "@/lib/navigation";
import { type ReactNode, useEffect, useState } from "react";

export default function Modal({ children }: { children: ReactNode }) {
  const [open, setOpen] = useState<boolean>(false);
  const router = useRouter();
  const pathname = usePathname();

  const handleClose = (open: boolean) => {
    if (!open) {
      setOpen(false);
      setTimeout(() => {
        router.back();
      }, 300);
    }
  };

  useEffect(() => {
    if (pathname === "/login" || pathname === "/cart") {
      setOpen(true);
    } else {
      setOpen(false);
    }
  }, [pathname, setOpen]);

  return (
    <Sheet open={open} onOpenChange={handleClose}>
      <SheetContent>
        <div className="h-full w-full overflow-auto">{children}</div>
      </SheetContent>
    </Sheet>
  );
}
```

**Comment by jiaweing** [comment]
This is my simple solution - check when the path changes using usePathname() and close it. The initialRender is to prevent it from closing immediately when the dialog opens.

This solution works for Dialog, Drawer and [Credenza](https://github.com/redpangilinan/credenza).
Hope this helps someone.

```js
  const pathname = usePathname()
  const initialRender = useRef(true)
  const [open, setOpen] = React.useState(false)

  useEffect(() => {
    if (initialRender.current) {
      initialRender.current = false
    } else {
      setOpen(false)
    }
  }, [pathname])

  return (
    <Dialog open={open} onOpenChange={setOpen}>
      <DialogContent>
        <DialogHeader>
          <DialogTitle>New Space</DialogTitle>
          <DialogDescription>
            Organise all your files and notes in one place.
          </DialogDescription>
        </DialogHeader>
        <CreateSpaceForm teamId={teamId} />
      </DialogContent>
    </Dialog>
  )
```

**Comment by ArturLinnik** [comment]
Still doesn't work in 14.2.4...

**Comment by gabrielcnegre** [comment]
So, I found a solution that seems to indicate our issue here might be an edge case not working as expected, my modal was not closing, when my tree was like this (page as layout sibbling):

- root/
- - layout.tsx
- - page
- - @modal

Now its working:

- root/
- layout.tsx
- - @children
- - - page.tsx
- - @modal

**Reply by kamiranoff** [reply]
> So, I found a solution that seems to indicate our issue here might be an edge case not working as expected, my modal was not closing, when my tree was like this (page as layout sibbling):
> 
> * root/
> * * layout.tsx
> * * page
> * * @modal
> 
> Now its working:
> 
> * root/
> * layout.tsx
> * * @children
> * * * page.tsx
> * * @modal

Not sure I understand the new structure...

I can see in the documentation:
> The children prop is an implicit slot that does not need to be mapped to a folder. This means app/page.js is equivalent to app/@children/page.js.

I don't understand in the new structure what goes inside children - which page?

**Reply by gabrielcnegre** [reply]
@kamiranoff The modal page. 

It's been some time now, but from what I remember is that in this specifc case I needed to use an explicit @children directory (contrary to the docs).

**Comment by kamalbennani** [comment]
Any update on this?

**Comment by MohamadrezaKazemian** [comment]
still facing this issue in 14.2.4

**Comment by eric-mathison** [comment]
Same issue in 14.2.15
It seems any parent route that is 'under' the layout providing the parallel routes will still show the route in the dom. Using nextjs's example for a login modal has this bug. Even if you have a catchall page, if you navigate to the parent route '/' from within the modal, the modal will continue to appear.

To add to this, if you have two modals in the same layout (ie. a login and register parallel & intercepted routes) if you link from one modal to the other route, BOTH modals will appear on the screen. 
For example, you have a link to your /register page on the login modal. 

I created a sandbox demonstrating this issue.
https://codesandbox.io/p/devbox/parallel-modal-test-wfqym4
![Snip - Preview - workspace - CodeSandbox - Google Chrome](https://github.com/user-attachments/assets/deececd3-50be-472b-bf47-06cf8251c876)

**Comment by tao-m10c** [comment]
15.0.3 and same behaviour.

**Reply by leonardolessa** [reply]
same on 15.1.4

**Reply by engelkes-finstreet** [reply]
Same for me! This is a really common pattern - a solution is highly appreciated

**Comment by renekahr** [comment]
15.1.6 still not working

**Comment by lamb572** [comment]
After having this issue, what seems to have worked for me was changing this step in the docs from `default.tsx` to `page.tsx` 
 

> <img width="668" alt="image" src="https://github.com/user-attachments/assets/9417ae2a-0233-4281-ae0d-b3949bcd73e7" />

**Reply by burak-kara** [reply]
Came here to write the same thing. Instead of trying the variants of catchall (e.g., `[...folderName] or [[...folderName]]`) as it is not allowed with a `page.tsx` file on the same level, just have a `page.tsx` file returning null.

**Reply by adriandlam** [reply]
Crazy how this one fix changed so many hours of searching. Kudos!

@rauchg please update docs?

**Comment by zinovyev** [comment]
Adding a `page.tsx` in the root of the slot fixed that for me. That's described in the official docs here: https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#closing-the-modal

So, what you need is the `page.tsx` returning `null` in the root of the `@authModal` I guess:

```
app
├─ @authModal
│  │ (.)auth
│  │  └─ login
│  │     └─ page.tsx // Contains the modal with login form 
│  │   
│  └─────── page.tsx // Returns null <-
```

**Reply by stupside** [reply]
Worked for me instead of having a default.ts

**Reply by byronncat** [reply]
Worked for me too

**Reply by melkir** [reply]
I would suggest keeping the `default.tsx` because it's useful when you're directly navigating to the URL.
e.g. `/` -> `/gallery` (open the modal)
Then if we refresh the page from `/gallery`, without the `default.tsx` you'll get a not found.

Keeping both `page.tsx` and `default.tsx` seems to works for me with the combination of replacing the `router.back()` by a `<Link>` component.

**Comment by mlecoq** [comment]
I'd like to have a `router.pop` feature, to avoid such ugly pattern:

```
 onClick={() => {
        const pop = searchParams.get('pop');
        if (pop && !isNaN(Number(pop))) {
          for (let i = 0; i < Number(pop); i++) {
            router.back();
          }
          return;
        }

        router.back();
      }}
```

**Comment by metafoo** [comment]
Still a problem in 2026 :(

After applying one hacky solution after another I finally ended up just sequentially performing `back()` and `push()`.

```
export function useModalRedirect() {
  const router = useRouter()

  return useCallback((path, inModal = false) => {
    if(!inModal) {
      router.push(path)

      return
    }

    function onPop() {
      window.removeEventListener("popstate", onPop)
      router.push(path)
    }

    window.addEventListener("popstate", onPop)
    router.back()
  }, [router])
}
```
